WIP / experiment: enable native unicode arr concatenation

tylerjereddy opened the pull request:
This is just an experiment that allows adding two unicode type arrays and getting an element-wise concatenated result without going through `numpy.core.defchararray.add()`.

So, `np.array(['a', 'b']) + np.array(['c', 'd']) -> np.array(['ac', 'bd'])`

The new unit test added does pass, but some other tests fail pretty badly for now.

I added some detailed comments to see if others think I'm misunderstanding here. Not suggesting we should do this, but more looking for feedback on what ideas I might have wrong re: making "string-type" handling better (which is on our roadmap).

tylerjereddy committed:
experiment: enable native unicode arr concat.

tylerjereddy committed:
BUG: prevent a subset of float64 reprs (inf-containing it seems) from calling array_add() of ndarray nb_add slot.

tylerjereddy commented:
The latest commit (5f41d33935e) improves the testing situation because of the issue outlined below:

This used to segfault on this branch:
```Python
print(np.array([np.nan, np.inf]))
```

Because `array_repr()` for that `float64` array actually calls `array_add()`, the function pointer in the `nb_add` slot of `ndarray` type struct itself, for a simple sum like: `3` + `False` to determine output padding. (where `3` is the len() of `inf` string and `False` is the bool val of `neginf` variable)

My observation of `repr` related segfault sensitivity here may relate to `repr` related crash sensitivity noted by @njsmith for `PyArray_GenericBinaryFunction()` in `number.c`, which is just downstream of `array_add()` in the control flow. I'm curious to know if systematic prevention of having array_repr() for various array "types" / combinations of elements call array_add() would be generally helpful?

tylerjereddy committed:
BUG: np.ndarray nb_add dispatch is now protected against case where self is not an array object.

tylerjereddy committed:
MAINT: clean up build warnings from unicodetype_concat().

tylerjereddy commented:
Amazingly, the CI is all green and the patch diff has 100 % coverage.

I suppose this is a simple example of "overloading" the function pointed to by the `ndarray` type struct `nb_add` slot with "multiple / specialized dispatch."

mattip commented on numpy/core/src/multiarray/scalartypes.c.src:
```diff
@@ -4255,12 +4255,23 @@ initialize_numeric_types(void)
+     * none of the functions pointed to here even exist; trying
+     * to create them has no effect on behavior -- they
+     * are currently circumvented
+     */
```
The functions exist, they come from `scalartypes.c.src`, line 474 and are generated.  When I run tests under a debugger, I hit `unicodetype_str` quite often in `test_arrayprint.py`

mattip commented on numpy/core/src/multiarray/number.c:
```diff
@@ -292,10 +299,109 @@ PyArray_GenericInplaceUnaryFunction(PyArrayObject *m1, PyObject *op)
+            PyArray_DESCR((PyArrayObject *)m2)->type_num == NPY_UNICODE) {
+            return unicodetype_concat((PyObject *)m1, m2);
+        }
+    }
```
What happens if `PyArray_Check(m2)` is false? For instance, try adding another statement in the new test: `test + u'abc'`

tylerjereddy committed:
Add support for unicode_ + scalar unicode.

tylerjereddy commented:
Ok, I added support for adding a "scalar" Python unicode object to a `unicode_` array. This only applies to direct usage of the `__add__` slot, so it will not likely be commutative yet.

That doesn't necessarily mean that anything else is supported either--I suppose I could start "overloading" other C functions to empower more slot operations if this is something we want to experment with further.

I probably don't gracefully handle shape mismatches beyond the scalar case yet either.

tylerjereddy committed:
MAINT: fix initialization warning for other_item.

tylerjereddy committed:
MAINT: handle unicode_ add shape mismatch

* unicodetype_concat() and array_add() now
handle shape mismatches by raising a ValueError
when two unicode_ arrays with incompatible
shapes use the __add__ slot of ndarray; a test
was added and this should restore 100 % patch
diff coverage as well

* rebased on latest master branch to alleviate
stochastic shippable test failures

tylerjereddy committed:
ENH: support scalar u'' + unicode

* add support for left-side operand Python
unicode object added to right-side operand
NumPy unicode_ array object; this is a bit
like mimicking support for __radd__

* added corresponding unit test

tylerjereddy committed:
TST: add an initial 2D unicode_ concat test and a broadcast concat test; the latter fails for now.

tylerjereddy committed:
ENH: unicode_ concat broadcasting

* added support for unicode_ concat
operation broadcasting between arrays
with different shapes

* now using PyArray_MultiIter
infrastructure to handle unicode_
concat array iteration

tylerjereddy commented:
Improved `unicode_` array concat operations to perform broadcasting -- apparently this isn't even possible with current `numpy.core.defchararray.add()`. Not sure if this is something we want to support, but not sure why not either.

tylerjereddy commented:
Closing based on discussion with @stefanv -- I guess this isn't quite aligned with priorities for time investment. If people disagree they'll probably have to speak up in favor of the direction so I don't get in trouble :)

I note that some potentially important core issues with segfaults on reprs noted by @njsmith in the source code might have been addressed by changes in this PR as discussed above and in the source comments, and that 50 % of the bullets on our current roadmap [for dtypes](https://www.numpy.org/neps/roadmap.html#better-dtypes) involve strings.

stefanv commented:
@tylerjereddy My hope was that you could still save the work in this PR that benefited NumPy; perhaps closing is premature—or are you aiming to add some of the fixes back in other PRs?

tylerjereddy commented:
This wasn't / isn't a priority, didn't use inner loops, and the alleged bug doesn't have a clear minimum reproducer available outside the PR, so closing this.
